### DISCUSSION TITLE
Using absurd-sql approach to support blazor wasm

### DISCUSSION THREAD

**Discussion by gentledepp** [ericsink/SQLitePCL.raw#440] [Ideas]
Hello @ericsink!

I recently came across [absurd-sql](https://github.com/jlongster/absurd-sql)
Essentially, this guy is using the emscripten transpiled version of sqlite in the browser.
Normally, this would be in-memory only, but he managed to tweak the sqlite code to use the IndexDb as a file store.

While this sounds **absurd** this is 10 times faster than indexdb and actually works!
You can read his in-depth explanations here: https://jlongster.com/future-sql-web 

Now, as SQLitePCL.raw is the foundation of all sqlite ORMs I know of (Microsoft.Data.SQLite => so even EF Core, sqlite-net-pcl, etc.) it would be awesome if there was a way to add support for absurt-sql somehow.

What I am thinking of is using Blazor WASM with full offline and persisted SQLite support.
This would be a game changer, as suddely, we could write completely offline capable apps for the browser - or even port them there. 
Moreover we could write PWAs and, thus, deliver mobile applications without ever touching any app store or play store.
This means: Faster deployment for custom apps to customers.

Do I sound like I have lost my mind, or does this kind of make sense to you?

Thank you for your time - any insights of you are very well appreciated!

**Comment by SteveSandersonMS** [comment]
@gentledepp See https://github.com/SteveSandersonMS/BlazeOrbital for a sample app that already does this.

**Reply by gentledepp** [reply]
Thank you so much for your link and the awesome sample app. I loved your presentation about this a while ago 😍

Maybe I do not understand the implementation well enough, but I disagree that "it already does this"

What absurd-sql does is to store sqlite pages in indexdb.

What blazeorbital [dbstorage.js](https://github.com/SteveSandersonMS/BlazeOrbital/blob/main/BlazeOrbital/ManufacturingHub/wwwroot/dbstorage.js) does is: every second (=> not transactional), in case the sqlite db file changed, store the whole file in indexdb.

1. it does not use pages as far as I see but stores a single file
2. it does so every second..... not ACID

Also, while I do see that entity framework magically can use e_sqlite3.o, I am not a native developer 😢so I do not quite understand **how** this was achieved.

What I would like to see is:
3. that you can use any library depending on SQLitePCL.raw to do ACID transactions on a local, persistent SQLite database in Blazor WebAssembly

But I don't see that. Could you please provide a link, or let me know how Blazeorbital magically wires **EF Core** to **e_sqlite3.o**? 

Thank you for your time - much appreciated!

**Reply by bricelam** [reply]
> ...how Blazeorbital magically wires EF Core to e_sqlite3.o

EF Core just calls SQLitePCLRaw APIs. See [the Providers section](https://github.com/ericsink/SQLitePCL.raw/blob/master/README.md#providers) of the README on this repo to see how SQLitePCLRaw loads the native library.

**Reply by gentledepp** [reply]
Ok, thanks. But then the question still remains: Is this a really usable approach for applications that need ACID write access to a local database?

To me it still looks like after some random interval (1 second) the whole sqlite database file is stored as a singe file in indexdb...

However, absurdsql stores the SQLite pages separately in indexdb. This sounds much more efficient to me as you do not have to write/read the **whole database** on every change.

Or am I missing something?

Could you please elaborate on that? This would be much appreciated.
A simple "you are right - this is just a sample app and **by no means** a usable approach for real world applications" would suffice though. Because I have the feeling it is...

**Reply by SteveSandersonMS** [reply]
It might help if you could describe in more detail what causes you to require server-level performance and durability guarantees on a client-side database. In most cases for things like mobile apps, the client-side db isn’t trusted to behave so robustly (eg it can always roll back if the user loses their device and restores a backup). I would expect a client-side DB commonly to be a local cache of data used for quick local queries and offline use. And in general it can’t be trusted not to be mutated arbitrarily by the user independently of your business rules. And the user may just delete it at any moment when clearing browser caches, and especially so if the size becomes large (eg hundreds of MB).

This is not to deny that other use cases exist! You absolutely may have legit needs for something else, but it would be great to clarify what and why so that we can build a picture of usage models and what approaches would be viable.

**Reply by SteveSandersonMS** [reply]
Also I should add: nobody is committed to the specific approach from BlazeOrbital of persisting the whole DB as a single blob. The details of how persistence works can be separated from most of the other concerns here so that developers can implement alternate strategies.

**Reply by gentledepp** [reply]
Cool! So my issue is I do like the approach of absurdsql but I am not a native developer (mostly C#, JS and typescript) so I have no clue how to do this unfortunately.

We would like to use this approach in an application for field agents that do inspections of buildings, power plants, etc. off grid. Currently we are using Xamarin and we do sync quite a lot of data using SQLite. Sometimes our customers have a local database with over 100MB.
There they perform their inspeciton and all defects or new equipment is added to the local db and synchronized to the central server whenever the client goes online again.

So loading 100mb every second or storing it would be way to slow. This is why I am so interested in the paging solition of absurdsql.

So: yeah it i still "just" a local cache, but we have to be able to mutate it efficiently and robustly nevertheless.

Does this make more sense now?
Do you have an idea/a colleague that could point out how "alternate strategies" could be implemented? (Did you or any of your colleagues ever have a look at absurdsql?)

**Reply by SteveSandersonMS** [reply]
> added to the local db and synchronized to the central server whenever the client goes online again. ... So loading 100mb every second or storing it would be way to slow

It's possible there's some misunderstanding of the data flow here. None of this has got anything to do with synchronizing to a central server over the network. I'd never recommend sending back the whole SQLite DB to a central server in any normal case; instead you'd just call server-side API endpoints that accept specific changes like "add new equipment item", and then the user would be able to resolve any conflicts interactively in the UI.

What happens in the BlazeOrbital case is only about copying from transient in-memory JS data to in-browser IndexedDB storage so that it's not lost if they reload the page. It's all local to the client, so completes almost instantly. And it doesn't happen every second - it only happens once within one second after the user actually writes a change to the DB, e.g., when they click a "Save" button on some form.

> Do you have an idea/a colleague that could point out how "alternate strategies" could be implemented?

Yes, you could change the code [here](https://github.com/SteveSandersonMS/BlazeOrbital/blob/main/BlazeOrbital/ManufacturingHub/wwwroot/dbstorage.js) to persist the state in any other way that you want. This doesn't invalidate your suggestion that an AbsurdSQL-like approach could reduce the copying by supplying further page-level data about what specifically has changed within this blob. There may be cases where people benefit from that, but it would be relatively rare that you'd be able to tell the difference, plus it creates new failure modes where perhaps only a subset of the pages have been persisted (such an issue can't happen in the BlazeOrbital strategy, where the whole DB file is written atomically).

**Reply by gentledepp** [reply]
thank you so much for the clarification.
Firstly - there is no misunderstanding. I did not expect that we send a full database to the server. But thanks for pointing this out.

However, you clearly do store the full database every 1 second given it has changed. So this might not be that much of a problem on a pc, but it definitely is on a smartphone.
Our goal would have been to move from Xamarin to a PWA in certain scenarios as it provides more flexibility. The key piece we were missing was a robust approach to handle a vast amount of relational data in an offline scenario. So I guess we will have to search for another approach then.
Thanks for pointing out the inconsistency issue with absurdsql, though!

**Reply by SteveSandersonMS** [reply]
> store the full database every 1 second given it has changed

Given it has changed *since the last time*, yes.  This may mean once in total per user action. But if your DB changes continually (eg constantly streaming in frequent upstream changes over a websocket) then I definitely agree this would be a poor strategy!

**Reply by gismofx** [reply]
@SteveSandersonMS @gentledepp I have a similar use case. Users are in the field and a solid internet connection isn't always guaranteed. So, the preferred mode is persisting a complete local copy of the database and when the internet connection is stable, perform a sync. If done daily, syncs are only 20-50 records typically. Still, the whole database could get quite large(appx 600 mb) which would be nice to persist as a SQLite file somewhere locally.

**Comment by bricelam** [comment]
See also https://github.com/ericsink/SQLitePCL.raw/issues/472
